diff --git a/sites/all/modules/contrib/logintoboggan/logintoboggan.module b/sites/all/modules/contrib/logintoboggan/logintoboggan.module
index 258e245..0a5edf0 100755
--- a/sites/all/modules/contrib/logintoboggan/logintoboggan.module
+++ b/sites/all/modules/contrib/logintoboggan/logintoboggan.module
@@ -1131,7 +1131,7 @@ function _logintoboggan_process_redirect($redirect, $account) {
   $fragment = isset($redirect['fragment']) ? $redirect['fragment'] : '';
   // @hack for comments stored in cookies in livecomments.module jghyde
   // Check is there's a cookie for nid
-  if (isset($_COOKIE['Drupal_visitor_nid']) && !empty($_COOKIE['Drupal_visitor_nid'])) {
+  if (isset($_COOKIE['Drupal_visitor_nid']) && !empty($_COOKIE['Drupal_visitor_nid']) && !user_is_anonymous()) {
     $redirect['path'] = 'node/' . $_COOKIE['Drupal_visitor_nid'];
     $fragment = 'comments';
     $redirect['query'] = array();
