diff --git a/sites/all/modules/contrib/radioactivity/radioactivity.module b/sites/all/modules/contrib/radioactivity/radioactivity.module
index c6cd940..e50c53c 100644
--- a/sites/all/modules/contrib/radioactivity/radioactivity.module
+++ b/sites/all/modules/contrib/radioactivity/radioactivity.module
@@ -291,6 +291,15 @@ function _radioactivity_update_energy($entity_type, $entity_id, $field_name, $la
   // Ensure that file_field_update() will not trigger additional usage.
   unset($entity->revision);
 
+  
+  //@HACK prevent the field_body from being deleted.
+  unset($entity->body);
+  foreach($entity as $key => $value){
+    if(strpos($key, 'field_') !== FALSE && $key != $field_name){
+      unset($entity->{$key}); 
+    }
+  }
+
   // Invoke the field presave and update hooks.
   field_attach_presave($entity_type, $entity);
   field_attach_update($entity_type, $entity);
diff --git a/sites/all/modules/custom/livelogin/livelogin.module b/sites/all/modules/custom/livelogin/livelogin.module
index 996ea57..f29f8c2 100644
--- a/sites/all/modules/custom/livelogin/livelogin.module
+++ b/sites/all/modules/custom/livelogin/livelogin.module
@@ -185,38 +185,3 @@ function livelogin_init() {
   // causing all jQuery and Javascript to break on the site.
   drupal_add_library('system', 'jquery.cookie');
 }
-
-/**
- * implements hook_comment_presave()
- *
-function livelogin_comment_presave($comment) {
-  // Remove leading & trailing spaces from the comment subject.
-  $comment->subject = trim($comment->subject);
-  /*
-function _radioactivity_update_energy(
-  $entity_type,
-  $bundle,
-  $field_name,
-  $language,
-  $entity_id,
-  $energy_delta,
-  $time,
-  $force = FALSE
-)
-
-  if (strpos($comment->node_type, 'node')) {
-    $list = explode('_', $comment->node_type);
-    $bundle = $list[2];
-    $nn = node_load($comment->nid);
-    $entity_type = 'node';
-    $field_name = 'field_radioactivity';
-    $language = LANGUAGE_NONE;
-    $entity_id = $comment->nid;
-    $energy_delta = 100;
-    _radioactivity_update_energy($entity_type, $bundle, $field_name, $language, $entity_id, $energy_delta, time(), FALSE);
-    // clear field caches just to be sure
-    field_cache_clear();
-  }
-  
-} */
-
